cardano-node/src/Cardano/Node/Run.hs

```diff
@@ -320,20 +323,46 @@ handleSimpleNode p trace nodeTracers npm onKernel = do
              producers' = case (List.lookup nodeid producersList) of
                             Just ps ->  ps
                             Nothing -> error $ "handleSimpleNode: own address "
-                                         <> show mockNodeAddress
+                                         <> show mockNodeAddr
                                          <> ", Node Id "
                                          <> show nodeid
                                          <> " not found in topology"
 
          traceWith tracer $ unlines
                                [ ""
                                , "**************************************"
-                               , "I am Node "        <> show mockNodeAddress <> " Id: " <> show nodeid
+                               , "I am Node "        <> show mockNodeAddr
+                                          <> " Id: " <> show nodeid
                                , "My producers are " <> show producers'
                                , "**************************************"
                                ]
 
-         when runDBValidation $ traceWith tracer "Performing DB validation"
+         when mockValidateDB $ traceWith tracer "Performing DB validation"
+
+-- | We provide an optional cross-platform method to politely request shut down.
+--
+-- The parent process passes us the file descriptor number of the read end of a
+-- pipe, via the CLI with @--shutdown-ipc FD@. If the write end gets closed,
+-- either deliberatly by the parent process or automatically because the parent
+-- process itself terminated, then we initiate a clean shutdown.
+--
+withShutdownHandler :: NodeProtocolMode -> Tracer IO String -> IO () -> IO ()
+withShutdownHandler (RealProtocolMode NodeCLI{shutdownIPC = Just (Fd fd)})
+                    tracer action =
+    race_ waitForEOF action
+  where
+    waitForEOF :: IO ()
+    waitForEOF = do
+      hnd <- IO.fdToHandle fd
+      r   <- try $ IO.hGetChar hnd
+      case r of
+        Left e | IO.isEOFError e -> traceWith tracer "received shutdown request"
+               | otherwise       -> throwIO e
+
+        Right _  ->
+          throwIO $ IO.userError "--shutdown-ipc FD does not expect input"
```

Comment:
Good idea. Reporting an error in this case is probably more sensible than consuming the input.